core/arch/riscv64/emit_utils.c

```diff
@@ -540,8 +574,28 @@ fragment_ibt_prefix_size(uint flags)
 void
 insert_fragment_prefix(dcontext_t *dcontext, fragment_t *f)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    /* Always use prefix on RISCV64 as there is no load to PC. */
+    byte *write_start = vmcode_get_writable_addr(f->start_pc);
+    byte *pc = write_start;
+    ASSERT(f->prefix_size == 0);
+
+    instrlist_t ilist;
+    instrlist_init(&ilist);
+
+    APP(&ilist,
+        INSTR_CREATE_ld(
+            dcontext, opnd_create_reg(DR_REG_A0),
+            opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0, TLS_REG0_SLOT, OPSZ_8)));
+    APP(&ilist,
+        INSTR_CREATE_ld(
+            dcontext, opnd_create_reg(DR_REG_A1),
+            opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0, TLS_REG1_SLOT, OPSZ_8)));
+
+    pc = instrlist_encode(dcontext, &ilist, pc, false);
+    instrlist_clear(dcontext, &ilist);
+
+    f->prefix_size = (byte)(((cache_pc)pc) - write_start);
```

Comment:
Both pc, and write_start are byte*, what's the reason to cast pc to cache_pc?

Reply:
I think it's copied from the other 3 arches... fixed.